TESTING.md

```diff
@@ -0,0 +1,87 @@
+# Testing Code Climate Engines
+
+The `codeclimate` CLI has a `test` command for automatically running a suite of regression tests that should be provided by the engine maintainer.
+
+The tests described in this document are designed to assert that the results of your engine are what you expect. For that reason, it is designed at a particular level of granularity that allows for detailed assertions to be made about the position, categorization, and content of the issues emitted by your engine.
+
+This document describes how to configure your engine to automatically run tests when `codeclimate test YOUR_ENGINE_NAME` is executed, describes the engine test format, and discusses integrating engine tests into your engine development workflow.
+
+## Configuring Your Engine To Run Tests
+
+Your engine can be configured to run tests by supplying path information in your engine's `engine.json` file:
+
+```
+{
+  "name": "fixme",
+  "description": "Discover lurking FIXMEs, BUGs, TODOs, etc. in your code",
+  "maintainer": {
+    "name": "Michael R. Bernstein",
+    "email": "mrb@codeclimate.com"
+  },
+  "languages": ["*"],
+  "test_paths": [
+    "/usr/src/app/test/integration/"
+  ]
+}
+```
+
+*Note: By convention, the `engine.json` file must be in the root directory of your container.*
+
+When `codeclimate test` is run, the `test_paths` are unpacked, and every file in the directory is run as a test (under the hood, we run `codeclimate analyze` with the `--engine` option and the test file path).
+
+Now that your engine can run tests, let's learn about how they're written.
+
+## Engine Test Format
+
+Engine tests are designed to allow you to assert specific attributes of issues emitted by your engine. The tests themselves are implemented as inline comments in "fixture" files containing source code intended to emit issues. Here's a sample:
+
+```
+# [issue] check_name=BUG description="BUG found" category="Bug Risk"
+# BUG: This should use double equals
```

Comment:
Using FIXME is confusing because it's a test for a comment, and the escaping issue. We should not use FIXME tests in examples, and use something like ESLint instead.